test/conformance/websocket_test.go

```diff
@@ -0,0 +1,147 @@
+// +build e2e
+
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conformance
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/serving/test"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	connectRetryInterval = 1 * time.Second
+	connectTimeout       = 6 * time.Minute
+)
+
+// connect attempts to establish websocket connection with the Service.
+// It will retry until reaching `connectTimeout` duration.
+func connect(logger *logging.BaseLogger, ingressIP string, domain string) (*websocket.Conn, error) {
+	u := url.URL{Scheme: "ws", Host: ingressIP, Path: "/"}
+	var conn *websocket.Conn
+	waitErr := wait.PollImmediate(connectRetryInterval, connectTimeout, func() (bool, error) {
+		logger.Infof("Connecting using websocket: url=%s, host=%s", u.String(), domain)
+		c, resp, err := websocket.DefaultDialer.Dial(u.String(), map[string][]string{"Host": []string{domain}})
+		if err == nil {
+			logger.Info("Connection established.")
+			conn = c
+			return true, nil
+		}
+		body := new(bytes.Buffer)
+		body.ReadFrom(resp.Body)
+		logger.Infof("Connection failed: %v.  Response=%+v, ResponseBody=%q", err, resp, body)
+		return false, nil
+	})
+	return conn, waitErr
+}
+
+// While we do have similar logic in knative/pkg, it is deeply buried inside
+// the SpoofClient which is very HTTP centric.
+//
+// TODO(tcnghia): Extract the GatewayIP logic out from SpoofClient, and reuse
+// that here.
+func getGatewayIP(kube *kubernetes.Clientset) (string, error) {
```

Comment:
Conformance needs to support different underlying implementations, so we need to find a way to do this without rummaging around in Istio bits.

Reply:
Maybe the right answer here is:
1. Move this to e2e for now (and check in, open an issue to move this to conformance and add to the Conformance project)
2. Change our e2e testing to set up `xip.io` in `config-domain` when the istio ingressgateway gets an IP
3. Kill the spoofing client and just use DNS